crates/mpz-garble-core/src/evaluator.rs

```diff
@@ -122,111 +115,204 @@ impl Evaluator {
+    pub fn evaluate_batched<'a>(
+        &'a mut self,
+        circ: &'a Circuit,
+        inputs: Vec<EncodedValue<state::Active>>,
+    ) -> Result<EncryptedGateBatchConsumer<'_, std::slice::Iter<'_, Gate>>, EvaluatorError> {
+        self.evaluate(circ, inputs).map(EncryptedGateBatchConsumer)
+    }
+}
+
+/// Consumer over the encrypted gates of a circuit.
+pub struct EncryptedGateConsumer<'a, I: Iterator> {
+    /// Cipher to use to encrypt the gates
+    cipher: &'static FixedKeyAes,
+    /// Buffer for the active labels.
+    labels: &'a mut [Label],
+    /// Iterator over the gates.
+    gates: I,
+    /// Circuit outputs.
+    outputs: &'a [BinaryRepr],
+    /// Current gate id.
+    gid: usize,
+    /// Hasher to use to hash the encrypted gates
+    hasher: Option<Hasher>,
+    /// Number of AND gates evaluated.
+    counter: usize,
+    /// Total number of AND gates in the circuit.
+    and_count: usize,
+    /// Whether the entire circuit has been garbled.
+    complete: bool,
+}
+
+impl<'a, I: Iterator> fmt::Debug for EncryptedGateConsumer<'a, I> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "EncryptedGateConsumer {{ .. }}")
+    }
+}
+
+impl<'a, I> EncryptedGateConsumer<'a, I>
+where
+    I: Iterator<Item = &'a Gate>,
+{
+    fn new(gates: I, outputs: &'a [BinaryRepr], labels: &'a mut [Label], and_count: usize) -> Self {
+        Self {
             cipher: &(*FIXED_KEY_AES),
-            circ,
-            active_labels,
-            pos: 0,
+            gates,
+            outputs,
+            labels,
             gid: 1,
+            hasher: None,
+            counter: 0,
+            and_count,
             complete: false,
-            hasher,
-        };
-
-        // If circuit has no AND gates we can evaluate it immediately for cheap
-        if ev.circ.and_count() == 0 {
-            ev.evaluate(std::iter::empty());
         }
+    }
 
-        Ok(ev)
+    /// Enables hashing of the encrypted gates.
+    pub fn enable_hasher(&mut self) {
+        self.hasher = Some(Hasher::new());
     }
 
-    /// Evaluates the next batch of encrypted gates.
+    /// Returns `true` if the evaluator wants more encrypted gates.
     #[inline]
-    pub fn evaluate<'a>(&mut self, mut encrypted_gates: impl Iterator<Item = &'a EncryptedGate>) {
-        let labels = &mut self.active_labels;
+    pub fn wants_gates(&self) -> bool {
+        self.counter != self.and_count
+    }
 
-        // Process gates until we run out of encrypted gates
-        while self.pos < self.circ.gates().len() {
-            match &self.circ.gates()[self.pos] {
-                Gate::Inv {
-                    x: node_x,
-                    z: node_z,
-                } => {
-                    let x = labels[node_x.id()].expect("feed should be initialized");
-                    labels[node_z.id()] = Some(x);
-                }
+    /// Evaluates the next gate in the circuit.
```

Comment:
This comment could be better and explain that this function evaluates several gates up to the next AND-gate including it. Currently it states that it evaluates the next gate in the circuit, which is not quite correct.